lib/core/src/Cardano/Wallet.hs

```diff
@@ -1235,377 +1213,160 @@ normalizeDelegationAddress s addr = do
+          where
+            amount = view #coin   bundle
+            assets = view #tokens bundle
+
+selectAssetsNoOutputs
+    :: forall ctx s k result.
         ( HasTransactionLayer k ctx
         , HasLogger WalletLog ctx
         , HasDBLayer s k ctx
         )
     => ctx
     -> WalletId
-    -> DelegationAction
-    -> ExceptT ErrSelectForDelegation IO CoinSelection
-selectCoinsForDelegation ctx wid action = do
-    dep <- fmap stakeKeyDeposit $
-        withExceptT ErrSelectForDelegationNoSuchWallet
-            $ readWalletProtocolParameters @ctx @s @k ctx wid
-
-    (utxo, _, txp, minUtxo) <- withExceptT ErrSelectForDelegationNoSuchWallet $
-        selectCoinsSetup @ctx @s @k ctx wid
-    selectCoinsForDelegationFromUTxO @_ @k ctx utxo txp minUtxo dep action
-
-selectCoinsForDelegationFromUTxO
-    :: forall ctx k.
-        ( HasTransactionLayer k ctx
-        , HasLogger WalletLog ctx
-        )
-    => ctx
-    -> W.UTxO
-    -> W.TxParameters
-    -> W.Coin
-    -> W.Coin
-    -> DelegationAction
-    -> ExceptT ErrSelectForDelegation IO CoinSelection
-selectCoinsForDelegationFromUTxO ctx utxo txp minUtxo dep action = do
-    let sel = initDelegationSelection tl dep action
-    let feePolicy = feeOpts tl (Just action) Nothing txp minUtxo sel
-    withExceptT ErrSelectForDelegationFee $ do
-        balancedSel <- adjustForFee feePolicy utxo sel
-        lift $ traceWith tr $ MsgDelegationCoinSelection balancedSel
-        pure balancedSel
-  where
-    tl = ctx ^. transactionLayer @k
-    tr = ctx ^. logger @WalletLog
-
--- | Estimate fee for 'selectCoinsForDelegation'.
-estimateFeeForDelegation
-    :: forall ctx s k.
+    -> TransactionCtx
+    -> (s -> SelectionResult TokenBundle -> result)
+    -> ExceptT ErrSelectAssets IO result
+selectAssetsNoOutputs ctx wid tx transform = do
+    -- NOTE:
+    -- Could be made nicer by allowing 'performSelection' to run with no target
+    -- outputs, but to satisfy a minimum Ada target.
+    --
+    -- To work-around this immediately, I am simply creating a dummy output of
+    -- exactly the required deposit amount, only to discard it on the final
+    -- result. The resulting selection will therefore have a delta that is at
+    -- least the size of the deposit (in practice, slightly bigger because this
+    -- extra outputs also increases the apparent minimum fee).
+    deposit <- calcMinimumDeposit @_ @s @k ctx wid
+    let dummyAddress = Address ""
+    let dummyOutput  = TxOut dummyAddress (TokenBundle.fromCoin deposit)
+    let outs = dummyOutput :| []
+    selectAssets @ctx @s @k ctx wid tx outs $ \s sel ->
+        transform s (sel { outputsCovered = mempty })
+
+-- | Selects assets from the wallet's UTxO to satisfy the requested outputs in
+-- the given transaction context. In case of success, returns the selection
+-- and its associated cost. That is, the cost is equal to the difference between
+-- inputs and outputs.
+selectAssets
```

Comment:
Wondered a bit about why `selectAssets` is not in `TransactionLayer` rather than `calcMinimumCoinValue`, `initSelectionCritereia`, `calcMinimumValue`.

It seems `readNextWithdrawal` needs separate access to `calcMinimumCost`. It is used from the Server.hs module, but the result is directly put back into `TransactionCtx`.

I suspect that logic could fit in `initSelectionCriteria`, such that Wallet.hs and Server.hs don't need to know about it.

Reply:
> Wondered a bit about why selectAssets is not in TransactionLayer 

The main reason is because `selectAssets` contains the glue code for pulling information from the database and running the selection itself. The `TransactionLayer` is completely agnostic to the database. So, one could argue that it suffices to make the UTxO and protocol parameters arguments of the `selectAssets` function to make it work; but that is basically what the `performSelection` function is already. Moreover, we'll still need to put the glue / effectful code  somewhere to prepare the arguments. 

Beside, having separate functions as `calcMinimumCoinValue`, `initSelectionCriteria` and `calcMinimumValue` allows in principle for easier testing since these functions cover a very specific sub-part of the logic. 

> It seems readNextWithdrawal needs separate access to calcMinimumCost. It is used from the Server.hs module, but the result is directly put back into TransactionCtx. I suspect that logic could fit in initSelectionCriteria, such that Wallet.hs and Server.hs don't need to know about it.

I am afraid that this won't work because the users are actually driving the request for withdrawal, not the server. This is why the server needs to know about it, because it depends on what the user has chosen. If we were systematically withdrawing from the account without asking the user, I think we could indeed move that logic to `initSelectionCriteria`. Yet at the moment, withdrawals are only set in the `TransactionCtx` if:

a) They are large enough
b) They have been explicitly requested by users.